src/components/AlchemyRow.vue

```diff
@@ -0,0 +1,176 @@
+import { Assets, Growth } from "~/composables/Utilities";
+import { Color , Alch, Bubble} from "~/composables/Alchemy";
+import bubblesData from "~/data/bubbles.json"
+
+
+
+export default defineComponent({
+  name: "AlchemyRow",
+  components: {
+    GameAsset,
+  },
+  props: {
+    group: {
+      required: true,
+      type: String as PropType<Color>,
+    },
+    discount: {
+      required: true,
+      type: Array as PropType<number[]>,
+    },
+    idx: {
+      required: true, 
+      type: Number,
+    }
+  },
+  emits: ["custom-change"],
+  methods: {
+    customChange (event: { target: { value: string; id: number; }; }) {
+      console.log("Handle leveling bubble input")
+      console.log("Only emit when id == 14 or (id == 6 and group == yellow)")
+      this.alchemy.upgrades[this.props.group][this.props.idx] = parseInt(event.target.value);
+      if (this.props.idx == 14 || (this.props.idx == 6 && this.props.group == "Yellow")) {
+        this.$emit("custom-change", event.target.value);
+        console.log(`Emitted [${15-event.target.id}] ${event.target.value}`);
+      }
+    }
+  },
+  setup(props) {
+    const state = useState();
+
+    const alchemy = computed({
+      get: () => state.value.alchemy,
+      set: (value) => (state.value.alchemy = value),
+    });
+   
+    // TODO:  This could maybe move to the customChange method
+    const handleAlchemyGoal = (ev: Event, color: Color, i: number) => {
+      let target = <HTMLInputElement>ev.target;
+      let val = target ? target.value : "0";
+      alchemy.value.goals[color][i] = parseInt(val);
+    };
+    
+
+    // Compute all materials needed for the wanted upgrades
+    const computeMaterials = () =>  {
+    
+      let bubble = bubblesData[props.group][props.idx];
+
+
+      // Discounts are stored as e.g. 70 (%), however calculation the form of 0.3
+      let discountCalc = props.discount.map((a) => {return (100-a)/100;});    
+      let levelNow = alchemy.value.upgrades[props.group][props.idx] ?? 0;
+      let levelGoal = alchemy.value.goals[props.group][props.idx] ?? 0;
+      let materialHistogram = [0,0,0,0];
+      // Sum all materials of the different levels together into the histogram
+      for (var i = levelNow; i < levelGoal; i++) {
+        let levelMulti = Math.pow(1.35 - (0.3 * i) / (50 + i), i);
+        bubble.Materials.forEach((m, y) => {
+          if(m.isLiquid) {
+            materialHistogram[y] += m.Amount + Math.floor(i/20);
+          } else {
+            materialHistogram[y] += Math.round(m.Amount*levelMulti*discountCalc[discountCalc.length -1]);
+          }
+        });
+      }
+      // Return the histogram combined with the material name and leave empty when no material was used
+      return materialHistogram.map(function(e, test) {
+        if (test < bubble.Materials.length) {
+          return [bubble.Materials[test].Name,e.toLocaleString()];
+        }
+        return ""
+      });
+    };
+
+    const effectChange = () => {
```

Comment:
Can we just call this function directly?

Reply:
Not sure if that would make the code better